From 95266cde0d2cf848ae9c5873bde8c94203c14696 Mon Sep 17 00:00:00 2001
From: Oliver Heger <oheger@apache.org>
Date: Sat, 2 Dec 2006 18:43:13 +0000
Subject: [PATCH] CompositeConfiguration now overrides clearPropertyDirect()
 instead of clearProperty(). This ensures that correct update events are
 generated. Fix for CONFIGURATION-241

git-svn-id: https://svn.apache.org/repos/asf/jakarta/commons/proper/configuration/trunk@481599 13f79535-47bb-0310-9956-ffa450edef68
---
 .../configuration/CompositeConfiguration.java |  2 +-
 .../TestCompositeConfiguration.java           | 62 ++++++++++++++++---
 xdocs/changes.xml                             |  4 ++
 3 files changed, 59 insertions(+), 9 deletions(-)

diff --git a/src/java/org/apache/commons/configuration/CompositeConfiguration.java b/src/java/org/apache/commons/configuration/CompositeConfiguration.java
index 0df59e88a0..22c4e008db 100644
--- a/src/java/org/apache/commons/configuration/CompositeConfiguration.java
+++ b/src/java/org/apache/commons/configuration/CompositeConfiguration.java
@@ -271,7 +271,7 @@ public boolean isEmpty()
     /**
      * {@inheritDoc}
      */
-    public void clearProperty(String key)
+    protected void clearPropertyDirect(String key)
     {
         for (Iterator i = configList.iterator(); i.hasNext();)
         {
diff --git a/src/test/org/apache/commons/configuration/TestCompositeConfiguration.java b/src/test/org/apache/commons/configuration/TestCompositeConfiguration.java
index a88a8074d3..16a240a2ae 100644
--- a/src/test/org/apache/commons/configuration/TestCompositeConfiguration.java
+++ b/src/test/org/apache/commons/configuration/TestCompositeConfiguration.java
@@ -440,7 +440,7 @@ public void testStringArrayInterpolation()
         assertEquals("2nd element", "foo.bar2", array[1]);
         assertEquals("3rd element", "foo.bar3", array[2]);
     }
-    
+
     /**
      * Tests whether global interpolation works with lists.
      */
@@ -515,15 +515,61 @@ public void testCloneNotSupported()
      */
     public void testCloneEventListener()
     {
-        cc.addConfigurationListener(new ConfigurationListener()
-        {
-            public void configurationChanged(ConfigurationEvent event)
-            {
-                // Just a dummy
-            }
-        });
+        cc.addConfigurationListener(new TestEventListenerImpl());
         CompositeConfiguration cc2 = (CompositeConfiguration) cc.clone();
         assertTrue("Listeners have been cloned", cc2
                 .getConfigurationListeners().isEmpty());
     }
+
+    /**
+     * Tests whether add property events are triggered.
+     */
+    public void testEventAddProperty()
+    {
+        TestEventListenerImpl l = new TestEventListenerImpl();
+        cc.addConfigurationListener(l);
+        cc.addProperty("test", "value");
+        assertEquals("No add events received", 2, l.eventCount);
+    }
+
+    /**
+     * Tests whether set property events are triggered.
+     */
+    public void testEventSetProperty()
+    {
+        TestEventListenerImpl l = new TestEventListenerImpl();
+        cc.addConfigurationListener(l);
+        cc.setProperty("test", "value");
+        assertEquals("No set events received", 2, l.eventCount);
+    }
+
+    /**
+     * Tests whether clear property events are triggered.
+     */
+    public void testEventClearProperty()
+    {
+        cc.addConfiguration(conf1);
+        assertTrue("Wrong value for property", cc
+                .getBoolean("configuration.loaded"));
+        TestEventListenerImpl l = new TestEventListenerImpl();
+        cc.addConfigurationListener(l);
+        cc.clearProperty("configuration.loaded");
+        assertFalse("Key still present", cc.containsKey("configuration.loaded"));
+        assertEquals("No clear events received", 2, l.eventCount);
+    }
+
+    /**
+     * A test configuration event listener that counts the number of received
+     * events. Used for testing the event facilities.
+     */
+    static class TestEventListenerImpl implements ConfigurationListener
+    {
+        /** The number of received events.*/
+        int eventCount;
+
+        public void configurationChanged(ConfigurationEvent event)
+        {
+            eventCount++;
+        }
+    }
 }
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index 70e2a1fc7e..e8e7e5e60a 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -23,6 +23,10 @@
 
   <body>
     <release version="1.4-dev" date="in SVN">
+      <action dev="oheger" type="update" issue="CONFIGURATION-241">
+        CompositeConfiguration.clearProperty() now generates the correct
+        update events.
+      </action>
       <action dev="oheger" type="update" issue="CONFIGURATION-242">
         The configuration returned by HierarchicalConfiguration.subset()
         performed variable interpolation only in the keys that belong to the
